### DISCUSSION TITLE
[Feature] back up XMP files with different naming

### DISCUSSION THREAD

**Discussion by ericswpark** [immich-app/immich#3407] [Feature Request]
### The feature

Allow the CLI tool to back up sidecar XMP files with different naming. Currently, the CLI tool looks for the following naming scheme:

```
IMG_0321.jpg
IMG_0321.jpg.xmp
```

With this FR, photos and sidecar files from e.g. Apple Photos exports can be imported as well. Example file naming format:

```
IMG_0321.jpg
IMG_0321.xmp
```

Related to #3383

### Platform

- [ ] Server
- [ ] Web
- [ ] Mobile

**Comment by thegranddesign** [comment]
We might also need a larger feature for iPhoto/Apple Photo compatibility.  For example, GPS coordinates are output like so:

```
<x:xmpmeta xmlns:x="adobe:ns:meta/" x:xmptk="XMP Core 5.4.0">
   <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
      <rdf:Description rdf:about=""
            xmlns:exif="http://ns.adobe.com/exif/1.0/"
            xmlns:photoshop="http://ns.adobe.com/photoshop/1.0/">
         <exif:GPSSpeedRef>K</exif:GPSSpeedRef>
         <exif:GPSSpeed>0.0</exif:GPSSpeed>
         <exif:GPSTimeStamp>2009-01-01T00:00:00Z</exif:GPSTimeStamp>
         <exif:GPSImgDirection>291.76978416999998</exif:GPSImgDirection>
         <exif:GPSLongitudeRef>W</exif:GPSLongitudeRef>
         <exif:GPSAltitudeRef>0</exif:GPSAltitudeRef>
         <exif:GPSLongitude>68</exif:GPSLongitude>
         <exif:GPSLatitude>60.5</exif:GPSLatitude>
         <exif:GPSLatitudeRef>N</exif:GPSLatitudeRef>
         <exif:GPSImgDirectionRef>T</exif:GPSImgDirectionRef>
         <exif:GPSAltitude>121</exif:GPSAltitude>
         <exif:GPSHPositioningError>0.0</exif:GPSHPositioningError>
      </rdf:Description>
   </rdf:RDF>
</x:xmpmeta>
```

Immich currently displays this as in [Russia](https://duckduckgo.com/?q=60.5%252C+68) because it doesn't take into account the `Ref` fields, however the correct location is [Akpatok Island](https://duckduckgo.com/?q=60.5+N%252C+68+W).

**Reply by ericswpark** [reply]
@thegranddesign can you file issues for these? Seems like a bug with XMP parsing. Especially since this discussion is now closed as the different naming problem was implemented in a PR.

**Reply by ericswpark** [reply]
Ah I see you've already filed it in #3540 which was since fixed. Thanks!

**Comment by thegranddesign** [comment]
Tags are also not working.  Tags show up in the format of:

```
<x:xmpmeta xmlns:x="adobe:ns:meta/" x:xmptk="XMP Core 5.4.0">
   <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
      <rdf:Description rdf:about=""
            xmlns:photoshop="http://ns.adobe.com/photoshop/1.0/"
            xmlns:dc="http://purl.org/dc/elements/1.1/">
         <dc:subject>
            <rdf:Seq>
               <rdf:li>Some Tag</rdf:li>
               <rdf:li>Some Other Tag</rdf:li>
            </rdf:Seq>
         </dc:subject>
      </rdf:Description>
   </rdf:RDF>
</x:xmpmeta>
```

It looks like Immich currently doesn't support tags but this would be a point of incompatibility.

**Comment by thegranddesign** [comment]
Date created within the Apple Photos XMP file is not being recognized when it's stored in the "photoshop" metadata.

```
<x:xmpmeta xmlns:x="adobe:ns:meta/" x:xmptk="XMP Core 5.4.0">
   <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">
      <rdf:Description rdf:about=""
            xmlns:photoshop="http://ns.adobe.com/photoshop/1.0/"
            xmlns:dc="http://purl.org/dc/elements/1.1/">
         <photoshop:DateCreated>1800-08-21T16:01:09-05:50:36</photoshop:DateCreated>
      </rdf:Description>
   </rdf:RDF>
</x:xmpmeta>
```

This imported photo was not tagged with 1800 but was instead tagged with the "create date" of the photo.

**Comment by harrismcc** [comment]
Any updates on this stuff? Proper Apple Photos XMP support would be a huge deal for me!

**Comment by ericswpark** [comment]
Think this part needs to be updated:

https://github.com/immich-app/immich/blob/885eba2b7c1fc19d5a819199b47df4923d9478b7/cli/src/cores/models/asset.ts#L35-L40

Just a note to myself; I have a final tomorrow so I can't take a shot at writing a patch yet.

**Comment by aisbergde** [comment]
the immich implementation of sidecar naming is uncommon and not compatible with other software. I just now realized, why "normal" sidecar files (abc.jpeg + abc.xmp) are not used and not associated with assets - it is just not jet implemented. That's why geo data in immich and outside does not match for some assets.

On the other side this an old feature request.

Now I don't know, what to do:
- keeping metadata only in sidecar files is not possible, until immich will not accept them
- embedded metadata into the assets is not possible for all assets (video, raw), and it needs to change the content of the assets. On the other side immich assuming that assets are read only and don't want to re-read them.

How to solve?

Is it really so complicated to just accept standard and more compatible naming patterns for sidecar files?

**Reply by alextran1502** [reply]
> Is it really so complicated to just accept standard and more compatible naming patterns for sidecar files?

I saw many of your posts have this similar type of question, so I want to answer it.

- You work in tech, so you understand that any technical change is challenging, regardless of how big or small. With Immich's big codebase, there are a lot of moving parts we have to carefully consider and plan before writing any code 😃 
- Each maintainer and contributor has their hit list tailored to each interest in the software, so this keeps things fun to work on without feeling like the actual "work" work but something the team enjoys doing in their free time. We don't have any roadmap ETA because of that philosophy 😄
- The features that you want, proposed would have a likelihood of 80% that we want it in Immich too, so eventually we will get to it 😉 

Now, to the sidecar, to be honest, before working on Immich, I and maybe some team members didn't know what a sidecar was, so it would be hard to know what the "standard" to adhere to. With this feature request, we will better understand how to work on it in the future.

**Reply by aisbergde** [reply]
I really value the big work and the fast progress. Asking feature request doesn't mean, that they should all be implemented. As you wrote, developers pick up, what they are interested. It's like to generate ideas and to hope, someone would like it. And to explain, why it could be useful to implement.

Unfortunately, I am not experiences in the used stack, otherwise I would try to start to take part in the development. Because I really like the idea of the project.

**Reply by alextran1502** [reply]
Yep, I understand. I am trying just to give you clear communication. I really appreciate you taking the time for feature requests, suggestions, and questions

**Comment by aisbergde** [comment]
I just want to explain, why it could be a good idea, to support both naming variants or why it could be useful to implement an option to switch. And what are possible issues when interacting with other software.

https://en.wikipedia.org/wiki/Extensible_Metadata_Platform
https://en.wikipedia.org/wiki/Sidecar_file

> In most cases the relationship between the source file and the sidecar file is based on the file name; sidecar files have the same base name as the source file, but with a different extension.

There is not a clear definition, not a "right" or "default" format, if the .xmp will replace the asset extension or will be just extended the extension.

The current immich implementation is not wrong, it is one possible implementation, it has pros and cons.

Some naming templates are better for a usage by different programs. In the digiKam docs 
https://docs.digikam.org/en/setup_application/metadata_settings.html
they wrote:

> Filename for the sidecars set to `filename.ext.xmp`. For example, `image1.dng` will have a sidecar file named `image1.dng.xmp`. With the option **Sidecar file names are compatible with commercial programs** digiKam will create the XMP sidecar files with a compatible file name (`image1.xmp`) used by many commercial programs.

The first option would be compatible with immich, but is not so compatible with other "commercial programs", the compatibility option is not compatible with immich.

How to solve?
- We could also try to use the other digiKam option, to make it compatible with immich.
- And it looks like, a good way is embedded metadata into the assets, where this is possible. immich can read them.

But how this should work, when immich stores their metadata changes into sidecar files (rotation, geo data, changed dates, whatever is planned)? It would be fine to use a flexible way, to make these changes compatible with other software. Otherwise, metadata in immich and in other software will not match. This should be avoided.

**Comment by ericswpark** [comment]
~~Implemented by #7813~~

EDIT: never mind, this doesn't fix the issue with the CLI tool.

**Reply by ericswpark** [reply]
@jrasm91 I see the CLI tool was switched to using the Immich SDK on #7962. Do you know if this uses the same codepath as #7813? Will uploading XMP files with different naming work without any changes to the CLI tool?

**Reply by jrasm91** [reply]
No, I think we need to make a code change there as well.

**Reply by kokulshan** [reply]
I now have an open PR for the CLI tool #8186

**Comment by ericswpark** [comment]
PR #8186 by @kokulshan (thank you!) was merged, closing discussion!
